## コンポーネント配下のコンテンツを任意の領域に描画する-ポータル
### ダイアログに実装
```
// PortalBasic.js
import { useState } from 'react';
import { createPortal } from 'react-dom';
import './PortalBasic.css';

export default function PortalBasic() {

  // ① useStateで状態管理
  const [show, setShow] = useState(false);
  const handleDialog = () => setShow(s => !s);

  return (
    <form>
      <button type="button" onClick={handleDialog}
        disabled={show}>
        ダイアログを表示
      </button>

      // ② ボタンとイベントハンドラ
      // ③ createPortalの使用
      {show && createPortal(
        <div className="dialog">
          <p>Portalで生成されたダイアログ</p>
          <button type="button" onClick={handleDialog}>
            閉じる
          </button>
        </div>,
        document.getElementById('dialog')
      )}
    </form>
  );
}
```
```
// PortalBasic.css
// ④ スタイル
.dialog {
  position: fixed;
  top: 100px;
  left: 100px;
  height: 100px;
  width: 250px;
  z-index: 99999;
  display: block;
  border: 1px solid black;
  padding: 5px;
  background-color: white;
  box-shadow: 5px;
}
```
```
// index.js
root.render(
  <>

    // ⑤ トリガー
    <div id="dialog"></div>
    <PortalBasic />
  </>
);
```
① useStateで状態管理
- useStateを使って、ダイアログの表示状態（show）を管理しています。
- showがtrueならダイアログが表示され、falseなら非表示です。

② ボタンとイベントハンドラ
- handleDialog関数は、ボタンがクリックされたときにshowの値をトグル（反転）させます。
- これでダイアログを開いたり閉じたりできます。

③ createPortalの使用
- createPortal関数は、第一引数にReact要素（ここではダイアログの内容）、第二引数にDOM要素（ここではid="dialog"のdiv）を受け取ります。
- これにより、ダイアログは#dialogに配置されます。

④ スタイル
- PortalBasic.cssでダイアログのスタイルを定義しています。
- position: fixed;によって、ダイアログは固定された位置に表示されます。

⑤ トリガー
- index.jsでPortalBasicコンポーネントとdiv id="dialog"を描画しています。
- このdivがcreatePortalで指定された場所です。
  
このように、createPortalはReactコンポーネントをDOMツリー外の任意の位置にレンダリングするために使います。これはモーダルダイアログやトースト通知など、ページ上の特定の位置にコンテンツを表示する際に便利です。

## コンポーネントで発生したエラーを処理する-Error Boundary
```
// ErrorRoot.js
import { ErrorBoundary } from 'react-error-boundary';
import ErrorThrow from './ErrorThrow';

export default function ErrorRoot() {
  return (
    <>
      <h3>Error Boundaryの基本</h3>

      // ①
      <ErrorBoundary
        // onError={err => alert(err.message)}
        fallback={<div>エラーが発生しました。</div>}
      >
        <ErrorThrow />
      </ErrorBoundary>
    </>
  );
}
```
```
// ErrorThrow.js
export default function ErrorThrow() {

  // ②
  throw new Error('Error is occured in MyApp.');
  return (
    <p>正しく実行されました。</p>
  );
}

```
①ErrorBoundaryの設定
- エラーが発生する可能性がある部分をErrorBoundaryタブで括る
- fallback属性にエラー発生時にコンテンツを指定する
  
②エラーをスローするコンポーネント 
- ErrorThrow コンポーネントは、意図的にエラーをスロー（throw）しています。
- これはエラーがどのようにハンドリングされるかを確認するためのものです。
- このエラーが発生すると、ErrorBoundary がそれをキャッチし、fallback で指定されたコンポーネントが表示されます。
- その結果、ユーザーには「エラーが発生しました。」と表示されます。

### より詳細なエラーメッセージを生成する
```
// ErrorRetryRoot.js
import { ErrorBoundary } from 'react-error-boundary';
import ErrorRetryThrow from './ErrorRetryThrow';
import ErrorFallback from './ErrorFallback';

export default function ErrorRetryRoot() {

  // ②エラー時に実行される処理
  const handleFallback = ({ error, resetErrorBoundary }) => {
    const handleClick = () => resetErrorBoundary();
    return (
      <div>
        <h4>以下のエラーが発生しました。</h4>

        // ③エラーメッセージの表示
        <p>{error.message}</p>
        <button type="button" onClick={handleClick}>
          Retry
        </button>
      </div>
    );
  };
  const handleReset = () => console.log('Retry!!');

  return (
    <>
      <h3>Error Boundaryの基本</h3>
      <ErrorBoundary
        onReset={handleReset}

        // ①fallbackRender属性
        fallbackRender={handleFallback}
        // FallbackComponent={ErrorFallback}
      >
        <ErrorRetryThrow />
      </ErrorBoundary>
    </>
  );
}
```
```
// ErrorRetryThrow.js
export default function ErrorRetryThrow() {
  if (Math.random() < 0.6) {
    throw new Error("Error is occured in MyApp.");
  }
  return <p>正しく実行されました。</p>;
}

```
①fallbackRender属性
- エラー時に表示すべきコンテンツを生成する関数（レンダラー）を指定
- エラーが発生した場合に呼び出される関数
- この関数が返すReact要素が画面に表示されます。この例ではhandleFallback関数が指定されています。
    
②エラー時に実行される処理
- handleFallback関数はエラーが発生した場合に実行されます。
- この関数はエラー情報とError Boundaryをリセットするための関数（resetErrorBoundary）を引数として受け取ります。
- この関数内でresetErrorBoundaryを呼び出すhandleClick関数も定義しています。
- これが"Retry"ボタンがクリックされたときに実行されます。

レンダラーは下記のようなプロパティを含んだオブジェクトを受け取る
    
| プロパティ | 概要 |
|:---|:---:|
|error |Error Boundary配下で発生したエラー情報 |
|resetErrorBoundary |Error Boundaryをリセットするための関数 |

③エラーメッセージの表示
- handleFallback関数内でエラーメッセージ（error.message）を表示しています。
- このメッセージはErrorBoundaryが捕捉したエラーから取得されます。
- "Retry"ボタン：このボタンがクリックされると、handleClick関数が実行され、Error Boundaryの状態がリセットされます。
- これにより、エラーが発生したコンポーネントが再度レンダリングされ、ユーザーに再試行のチャンスが与えられます。
